design/continuous-delivery.md

```diff
@@ -0,0 +1,332 @@
+environment and set-up resources needed for deployment such as the assets S3
+bucket.
+
+Prototype: https://github.com/eladb/cdkcd-test
+
+## Requirements
+
+This list describes only the minimal set of requirements from this feature.
+After we release these building blocks, we will look into vending higher-level
+"one liner" APIs that will make it very easy to get started.
+
+1. **Assets**: Support apps that include all supported kinds of assets (S3
+   files, ECR images)
+2. **Multiple-environments**: Support apps that have stacks that target multiple
+   environments (accounts/regions)
+3. **Orchestration**: Allow developers to express complex deployment
+   orchestration based on the capabilities of CodePipeline
+4. **User defined build + synth environment**: the runtime environment in which
+   the code is built and the CDK app is synthesized should be fully customizable
+   by the user
+5. **Controlled deployment environment**: for security reasons, the runtime
+   environment in which deployment is executed will be fully controlled and will
+   not allow running user code or user-defined image
+
+_Non-requirements/assumptions:_
+
+1. Bundling assets (creating Lambda zip files, building docker images) and
+   publishing them (upload to S3, push to ECR) is currently coupled with the
+   deployment of each stack that uses them. This means, for example, that if a
+   docker image is used in 3 stacks, it will be built 3 times, and this could
+   lead to different images (for example, “apt get install” can install a
+   different version of a dependency each time). We are assuming that this is
+   good enough for the first version of this feature.
+2. We assume that **cdk.context.json** is committed into the repo. Any
+   context-fetching will be done manually by users and committed to the
+   repository.
+3. There’s a one-to-one mapping between an app and a pipeline. We are not
+   optimizing the experience for multiple apps per pipeline (although
+   technically it should be possible to do it, but it’s not a use case we are
+   focused on).
+4. We are not optimizing this experience to support any CD tool (e.g. Jenkins)
+   but we should provide guidance on how to use “cdk synth” and “cdk deploy” so
+   that users will be able to implement CI/CD through their tools.
+5. Dependency management, repository and project structure are out of scope: we
+   don’t want to be opinionated about how users should structure their projects.
+   They should be able to use the tools they are familiar with and are available
+   to their teams to structure and modularize their applications.
+
+## Approach
+
+At a high-level, we will model the deployment process of a CDK app as follows:
+**source** => **build** => **pipeline => deploy:**
+
+1. In the *source* stage the code is pulled from a source repository (e.g.
+   CodeCommit, GitHub or S3), like any other app.
+2. In the *build* stage, the application is compiled (if required) **and
```

Comment:
I like the premise of the cloud-assembly approach as a manner to structure a project or application, though some aspects seem contrary to an "Anything-as-Code" model. For example (in my oversimplified view of the world), If I create an EC2 instance, all I need are a set of parameters and I can functionally create/recreate/duplicate exactly the same thing every time, whether via the console, API or CloudFormation. If I understand the design choice of CDK, I now need to track and manage unique instances of CloudFormation templates and not just the code to generate the template as artifacts created by CDK are not idempotent. So rather than having a single repo that I can use to deploy 1 to N number of stacks, it looks like I need maintain N copies of the repo in order to manage N-stacks. Is this the case or am I out in left field on this one. This is not the case with tools such as Troposphere and Sceptre. Also how well does CDK fit with the AWS Service Catalog?

Reply:
Cloud assemblies are not about structuring your application, they are a build artifact produced from your CDK application. You can model any number of stacks deployed to any number of accounts/region within a single CDK app and during build, the synthesized cloud assembly will include a template for each stack.